feat: add option to store files

leonardovillela opened the pull request:
# Description
This PR introduces changes to support stores in the host machine files uploaded. The new option `storeFiles` with default value as `true` determines it, if `true` the files are parsed and stored, if `false` the files are only parsed. Closes #609.

# Key Points
- I tried to maintain this retro compatible and not introduce any breaking change.
- I thought to have a class for the base file to avoid duplication of code between `VolatileFile` and `PersistentFile`, but in the end, I thought it was not worthy. Let me know what you think about it.
- I focused more on integration tests for this PR because I think it maps better for this type of change, but let me know if I need to add more unit tests.

GrosSacASac commented:
Could you make an example that show how to use volatile file as an input to another http request body ?

tunnckoCore reviewed (dismissed):
Haha, seems like a clever approach.

The only thing that bugs me every time and constantly forgetting _"what's the difference between `filename` and `name`, what's the diff between `type` and `mime`? Is there a difference?"_. I think we should get rid of them, only `filename` and `mime` to remain? But that's probably for later, still for v2 tho.

tunnckoCore commented on README.md:
```diff
@@ -334,6 +334,8 @@ See it's defaults in [src/Formidable.js DEFAULT_OPTIONS](./src/Formidable.js) (t
   for incoming files, set this to some hash algorithm, see
   [crypto.createHash](https://nodejs.org/api/crypto.html#crypto_crypto_createhash_algorithm_options)
   for available algorithms
+- `options.storeFiles` **{boolean}** - default `true`; to store uploaded file(s)
```
Probably add a hint in the `features` section where we say "automatically writes to disk (optional soon)", something like "see `options.storeFiles`?

leonardovillela commented:
> Could you make an example that shows how to use volatile file as an input to another http request body ?
@GrosSacASac not sure if I understand correctly what to do, it's to me add on README an example of usage of formidable to only parse files and store It?

leonardovillela commented:
> The only thing that bugs me every time and constantly forgetting "what's the difference between filename and name, what's the diff between type and mime? Is there a difference?". I think we should get rid of them, only filename and mime to remain? But that's probably for later, still for v2 tho.
@tunnckoCore I think they all represent the same things, so there is no difference. But I think we should not change the code to have the same nomenclature in all places, because in the context of an HTTP request the `mime` and `filename` nomenclature makes more sense than `type` and `name`, and in the context of file `mime` don't make sense because it's related to HTTP request context(a.k.a domain) and the `filename` sound a bit repetitive because we already in the context of a file.

GrosSacASac commented:
Readme or in examples https://github.com/node-formidable/formidable/tree/master/examples

leonardovillela commented:
@GrosSacASac and @tunnckoCore request changes are done :)

GrosSacASac commented:
How do you redirect the stream to  something else other than a file ? With volatilefile  ? How to overwrite the write () {} ? Or did I miss something.
The example does not show how.

tunnckoCore commented:
> How do you redirect the stream to something else other than a file ? With volatilefile ? How to overwrite the write () {} ? Or did I miss something.
> The example does not show how.

Not sure what you mean.

You can get the file from `{ files }` then redirect it wherever you want? That's what I understand, i think :thinking:

GrosSacASac commented:
`{ files } ` only contain VolatileFile instances, which do not contain the file content itself (write method never saves it).

GrosSacASac commented:
To be clear: There are 2 thing this PR should do:
 - An option to not store file in the fs
 - An option to do something else with the file

{ files } only contains file size and file names.

May be add another option besides storeFiles 
 - handleFileStream(name, stream)

GrosSacASac commented:
Or we can merge this now, and make another PR for part 2

tunnckoCore commented:
Oh, I got it. Fair point.

I think it would be better as part of this PR... I don't know.

Merry Christmas :wink:

leonardovillela commented:
Hey, @tunnckoCore and @GrosSacASac sorry for the late response. 

I added changes to write the file to something else(cloud storages, stdout and etc...) I also added two examples of how to do this. I opted to maintain only one option because I don't see much value in having two options that are coupled.

BTW the integration test is broken on CI, but on my machine, they are passing, I tried to fix this but I can't 😞 , can you help me?

GrosSacASac reviewed (approved):
Well done

Benjy96 commented:
Was this functionality removed?

GrosSacASac commented:
No
